ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/TransactionContractKeysValidation.scala

```diff
@@ -0,0 +1,295 @@
+  private case object Duplicate extends KeyValidationError
+  private case object Inconsistent extends KeyValidationError
+
+  private case class KeyValidationState(
+      activeDamlStateKeys: Set[DamlStateKey],
+      submittedDamlContractKeysToContractIds: Map[DamlContractKey, Option[RawContractId]] =
+        Map.empty,
+  ) {
+    def addSubmittedDamlContractKeyToContractIdIfEmpty(
+        submittedDamlContractKey: DamlContractKey,
+        contractId: Option[ContractId],
+    ): KeyValidationState =
+      copy(submittedDamlContractKeysToContractIds =
+        if (!submittedDamlContractKeysToContractIds.contains(submittedDamlContractKey)) {
+          submittedDamlContractKeysToContractIds + (submittedDamlContractKey -> contractId.map(
+            _.coid
+          ))
+        } else {
+          submittedDamlContractKeysToContractIds
+        }
+      )
+  }
+
+  private type KeyValidationStatus = Either[KeyValidationError, KeyValidationState]
+
+  private def validateTransactionContractKeysUniquenessAndConsistency(
+      transactionCommitter: TransactionCommitter,
+      recordTime: Option[Timestamp],
+      transactionEntry: DamlTransactionEntrySummary,
+      contractKeyDamlStateKeys: Set[DamlStateKey],
+      contractKeysToContractIds: Map[DamlContractKey, RawContractId],
+  ): StepResult[DamlTransactionEntrySummary] = {
+
+    val keysValidationOutcome = transactionEntry.transaction
+      .foldInExecutionOrder[KeyValidationStatus](
+        Right(KeyValidationState(activeDamlStateKeys = contractKeyDamlStateKeys))
+      )(
+        (keyValidationStatus, _, exerciseBeginNode) =>
+          validateNodeContractKey(
+            exerciseBeginNode,
+            contractKeysToContractIds,
+            keyValidationStatus,
+          ),
+        (keyValidationStatus, _, leafNode) =>
+          validateNodeContractKey(leafNode, contractKeysToContractIds, keyValidationStatus),
+        (accum, _, _) => accum,
+      )
+
+    keysValidationOutcome match {
+      case Right(_) =>
+        StepContinue(transactionEntry)
+      case Left(error) =>
+        val message = error match {
+          case Duplicate => "DuplicateKeys: at least one contract key is not unique"
+          case Inconsistent =>
+            "InconsistentKeys: at least one contract key has changed since the submission"
+        }
+        transactionCommitter.reject(
+          recordTime,
+          transactionCommitter.buildRejectionLogEntry(
+            transactionEntry,
+            RejectionReason.Inconsistent(message),
+          ),
+        )
+    }
+  }
+
+  private def validateNodeContractKey(
+      node: Node.GenNode[NodeId, ContractId],
+      contractKeysToContractIds: Map[DamlContractKey, RawContractId],
+      keyValidationStatus: KeyValidationStatus,
+  ): KeyValidationStatus =
+    validateNodeKeyConsistency(
+      contractKeysToContractIds,
+      node,
+      validateNodeKeyUniqueness(node, keyValidationStatus),
+    )
```

Comment:
I would rather you extracted the `flatMap` up several levels, to decouple uniqueness checks from consistency checks as much as possible. They don't actually seem to be related apart from aggregating into the same class.

Reply:
I don't want to traverse the transaction twice though. I'll see what I can do.

Reply:
Makes sense. I was wondering if you could also push the `fold` up, but maybe it won't work.

Reply:
I am not sure, as the monotonicity check doesn't need to fold, but I'll have a look.